Vhdx: Vhdx implementation, integration, and test

fazlamehrab opened the pull request:
This PR introduces the support for both fixed and dynamic VHDx file format, along with proper integration testing and fuzzing.

Fixes #2282

likebreath commented on tests/integration.rs:
From the log [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2888/checks?check_run_id=3128021107), you see the error is from the `Clippy (integration tests)` step with the following command from the GH workflow, and you can rerun the same command on your dev machine to reproduce the error:

```
Run cargo clippy --all --all-targets --tests --features "integration_tests" -- -D warnings
  cargo clippy --all --all-targets --tests --features "integration_tests" -- -D warnings
```

Also, for huge changes, I normally do a PR to my own fork to test those GH actions (which will avoid running the complete CI). Of course, you can manually do all of them by following the workflow defined in the repo, say [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/.github/workflows/quality.yaml#L33-L61).

fazlamehrab replied:
Thanks for the tips. It will help make a successful PR.

likebreath reviewed (commented):
@fazlamehrab Great to see the (huge) PR. Few simple tips on debugging/reproducing the GH action failures. Hope that helps.

fazlamehrab commented on tests/integration.rs:
```diff
@@ -2527,6 +2531,31 @@ mod tests {
+            let mut raw_file_path = workload_path.clone();
+            let mut vhd_file_path = workload_path;
+            raw_file_path.push(FOCAL_IMAGE_NAME);
+            vhd_file_path.push(FOCAL_IMAGE_NAME_VHD);
```
This should be VHDx. Thanks for pointing this out.

sboeuf commented:
@fazlamehrab there's an error with the DCO check for each commit:
```
Commit sha: b50d565, Author: fazlamehrab, Committer: fazlamehrab; Expected "fazlamehrab mehrab@vt.edu", but got "Fazla Mehrab akm.fazla.mehrab@intel.com".
```
That's because you've registered to Git with `fazlamehrab mehrab@vt.edu`. Make sure you modify this so that your default signoff is the one you expect and therefore the one DCO expects as well.

sboeuf commented on vhdx/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+
+[dependencies]
+byteorder = "1.4.3"
+libc = "0.2.94"
```
`0.2.98`

sboeuf commented on vhdx/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+remain = "0.2.2"
+vmm-sys-util = ">=0.3.1"
+crc32c = "0.6.0"
+uuid = { version = "0.8", features = ["serde", "v4"] }
```
I don't think you need the `serde` feature for this crate, do you?
The latest version is `0.8.2`.

sboeuf commented on Cargo.toml:
```diff
@@ -86,6 +86,7 @@ members = [
     "vm-migration",
-    "vm-virtio"
+    "vm-virtio",
+    "vhdx"
```
Add this alphabetically.

sboeuf commented on block_util/Cargo.toml:
```diff
@@ -13,6 +13,7 @@ io-uring = ">=0.4.0"
 libc = "0.2.98"
 log = "0.4.14"
 qcow = { path = "../qcow" }
+vhdx = { path = "../vhdx" }
```
Add this alphabetically.

sboeuf commented on block_util/src/lib.rs:
```diff
@@ -18,8 +18,10 @@ pub mod qcow_sync;
 pub mod raw_async;
 pub mod raw_sync;
 pub mod vhd;
+pub mod vhdx_sync;
```
Just a comment about your commit message. Please replace `vhdx: vhdx crate integration with the cloud hypervisor` with `block_util, vmm: vhdx crate integration with the cloud hypervisor`.

sboeuf commented on vhdx/src/lib.rs:
```diff
@@ -0,0 +1,5 @@
+pub mod vhdx;
```
Missing the Copyright section on top:
```
// Copyright © 2021 Intel Corporation
//
// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
```

sboeuf commented on vhdx/src/vhdx.rs:
```diff
@@ -0,0 +1,224 @@
+use crate::vhdx_bat::{BatEntry, Error as VhdxBatError};
```
Missing Copyright.

sboeuf commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,122 @@
+use crate::vhdx_header::RegionTableEntry;
```
Missing Copyright

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,518 @@
+extern crate log;
```
Ditto about Copyright

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,232 @@
+use crate::vhdx_bat::{self, BatEntry};
```
Ditto about Copyright.

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,319 @@
+use crate::vhdx_header::RegionTableEntry;
```
Ditto about copyright.

sboeuf commented:
@fazlamehrab I did a first pass and also sent an extra commit as part of this PR. You need to fix the Cargo.lock as part of the first or second commit, and then you also need to fix the integration test with the suggested changes.
I'll do a second pass over the vhdx crate content later today or tomorrow.

sboeuf commented:
I'll try to understand why the `aarch64` CI is failing, but otherwise you can see the CI is green :)

rbradford commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,232 @@
+                    let mut new_size =
+                        round_up!(disk_spec.image_size, vhdx_metadata::BLOCK_SIZE_MIN as u64);
+                    new_size += disk_spec.block_size as u64;
+                    if new_size > size_of::<u64>() as u64 {
```
I don't think you meant to write this. This just compares `new_size` to 8 (size of u64)

fazlamehrab replied:
Yes, you are right! Fixed.

rbradford commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,126 @@
+
+#[derive(Default, Clone, Debug)]
+pub struct BatEntry {
+    pub bat_entry: u64,
```
Why do you need this?

fazlamehrab replied:
This represents the BAT Entry structure described in the spec.

sboeuf replied:
You can use the following notation:
```rust
pub struct BatEntry(u64);
```
And then you would access the value from the code the following way:
```rust
let my_bat_entry = BatEntry(10);
println!("my_bat_entry value: {}", my_bat_entry.0)
```

fazlamehrab replied:
Why initializing with 10? Are you suggesting to replace the vector that contains bat entries as well?

sboeuf replied:
This was just an example of how to use the syntax :)

fazlamehrab replied:
Got it!

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,522 @@
+
+const REGION_ENTRY_REQUIRED: u32 = 1;
+
+const BAT_GUID: u128 = 11023197692918435611014362394968356710; // BAT GUID in integer read in little endian
```
You added the uuid crate, why not use it?

fazlamehrab replied:
Replied in the other other comment on uuid.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,323 @@
+const METADATA_VIRTUAL_DISK_SIZE: u128 = 245846085492271402621565837099086987812; //"2FA54224-CD1B-4876-B211-5DBED83BF4B8";
+const METADATA_VIRTUAL_DISK_ID: u128 = 94079244529923181982015449247999726251; //"BECA12AB-B2E6-4523-93EF-C309E000C746";
+const METADATA_LOGICAL_SECTOR_SIZE: u128 = 127169626291185313037823917095022346013; //"8141BF1D-A96F-4709-BA47-F233A8FAAB5F";
+const METADATA_PHYSICAL_SECTOR_SIZE: u128 = 115338139532893941197488005837198936263; //"CDA348C7-445D-4471-9CC9-E9885251C556";
```
Ditto about uuid crate.

fazlamehrab replied:
There are two problems. 
1. According to the spec, we are supposed to read a UUID as 128bit integer in little-endian format. However, this uuid crate converts UUIDs to 128 bit integers in big-endian format, which gives different values. Therefore, I manually figured the corresponding little-endian values for the UUIDs and hard coded here.

2.  We have a match statement where check if the parsed integer value matches one of the known UUIDs. Match statement only takes constants for patterns. If we had used uuid crate to convert the UUIDs to integer, then we would get variables, which are not acceptable in match. Therefore, we realized it is easier and simpler to hard code the UUID values in integers while formatted in little-endian.

fazlamehrab replied:
We using uuid crate now to convert from UUID to u128.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,323 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
```
Ditto.

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,522 @@
+    fn current_header(
+        header_1: Result<Header>,
+        header_2: Result<Header>,
+    ) -> (u8, Result<Header>, Result<Header>) {
```
Probably should be `Result<(u8, Header, Header)>`

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,522 @@
+        }
+
+        if no_valid_header {
+            panic!("Get current header: {}", Error::NoValidHeader);
```
Bad panic.

sboeuf commented on vhdx/src/vhdx.rs:
```diff
@@ -0,0 +1,228 @@
+            Ok(r) => Ok(r),
+            Err(e) => {
+                println!("Vhdx write: {}", e);
+                Err(io::Error::last_os_error())
```
You want to use `map_err()?` as well here in order to avoid losing the error you got from `vhdx_io::write()`.

sboeuf reviewed (commented):
Your signoff is wrong since you now have both `Signed-off-by: Fazla Mehrab <akm.fazla.mehrab@intel.com>` and `Signed-off-by: fazlamehrab <mehrab@vt.edu>`.
I think the one you want is `Signed-off-by: Fazla Mehrab <akm.fazla.mehrab@intel.com>` but it needs to be part of your global signature, otherwise DCO will complain.

sboeuf commented on vhdx/src/vhdx.rs:
```diff
@@ -0,0 +1,228 @@
+                return Ok(o);
+            }
+        }
+        Err(std::io::Error::from_raw_os_error(EINVAL))
```
Use https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.InvalidData instead.

fazlamehrab replied:
The Seek trait expects std::io::Error from seek(), where InvalidData is std::io::ErrorKind.

sboeuf replied:
Of course it expects `std::io::Error`, I was suggesting to create a new `std::io::Error` with the `ErrorKind` `InvalidData`.
```suggestion
        Err(std::io::Error::new(std::io::ErrorKind::InvalidData, format!("Failed seeking to {}", pos)))
```

fazlamehrab replied:
Thanks for the solution. Is it ok if we do not print pos? It is not an integer but a SeekFrom structure.

rbradford commented:
I think this needs a fuzzer. @likebreath can help you with that if you have trouble.

fazlamehrab commented on vhdx/src/vhdx.rs:
```diff
@@ -0,0 +1,228 @@
+    match FileTypeIdentifier::new(f) {
+        Ok(_) => Ok(true),
+        Err(_) => Ok(false),
+    }
```
Applied this. Could not think of any better solution yet.

rbradford commented on block_util/src/vhdx_sync.rs:
```diff
@@ -0,0 +1,116 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
```
You resolved it but didn't fix it?

fazlamehrab replied:
Just pushed my changes.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+
+struct MetadataTableHeader {
+    _signature: u64,
+    _reserved: u16,
```
No point including reserved fields if you're going to read it field by field vs using `#[repr(packed)]` and reading the whole thing in.

sboeuf replied:
I think he needs it for when he calls into `size_of::<MetadataTableHeader>()`.

fazlamehrab commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+    pub fn new(
+        disk_spec: &DiskSpec,
+        bat: &[BatEntry],
+        sector_index: &u64,
```
I don't remember exactly at what situation I did that, but we obviously don't a reference now. Thanks.

sboeuf commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,126 @@
+            f.seek(SeekFrom::Start(offset + i * size_of::<u64>() as u64))
+                .map_err(Error::ReadBat)?;
+
+            let bat_entry = f.read_u64::<LittleEndian>().map_err(Error::ReadBat)?;
+            let bat_entry = BatEntry(bat_entry);
```
Simplify like that:
```suggestion
            let bat_entry = BatEntry(f.read_u64::<LittleEndian>().map_err(Error::ReadBat)?);
```

sboeuf commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,126 @@
+        bat_offset: u64,
+        bat_entries: Vec<BatEntry>,
+    ) -> Result<()> {
+        let entry_count = bat_entries.len() as u64;
+        let offset = bat_offset;
```
You don't need to create some intermediate variables here, just use them directly in the loop.

sboeuf commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,126 @@
+    pub fn write_bat_entries(
+        f: &mut File,
+        bat_offset: u64,
+        bat_entries: Vec<BatEntry>,
```
Pass a reference:
```suggestion
        bat_entries: &[BatEntry],
```

sboeuf commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,126 @@
+            f.seek(SeekFrom::Start(offset + i * size_of::<u64>() as u64))
+                .map_err(Error::ReadBat)?;
+            let bat_entry = match bat_entries.get(i as usize) {
+                Some(entry) => (*entry).0,
```
```suggestion
                Some(entry) => entry.0,
```

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+    }
+}
+
+macro_rules! round_up {
```
I'd call this `align` instead of `round_up` since that's what it does.

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+            InvalidBatEntryState => write!(f, "Invalid BAT entry state"),
+            InvalidBatIndex => write!(f, "Invalid BAT entry count"),
+            InvalidDiskSize => write!(f, "Invalid disk size"),
+            ReadSectorBlock(e) => write!(f, "Failed reading sector blocks form file {}", e),
```
```suggestion
            ReadSectorBlock(e) => write!(f, "Failed reading sector blocks from file {}", e),
```

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+    pub fn new(
+        disk_spec: &DiskSpec,
+        bat: &[BatEntry],
+        sector_index: &u64,
+        sector_count: &u64,
```
```suggestion
        sector_index: u64,
        sector_count: u64,
```

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+        sector.bat_index = *sector_index / disk_spec.sectors_per_block as u64;
+        sector.block_offset = *sector_index % disk_spec.sectors_per_block as u64;
+        sector.free_sectors = disk_spec.sectors_per_block as u64 - sector.block_offset;
+        if sector.free_sectors > *sector_count {
```
Shouldn't this be an error? I mean in case the calculated `free_sectors` is bigger than what is expected, that sounds like we should throw an error.

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+        sector.block_offset *= disk_spec.logical_sector_size as u64;
+
+        let bat_entry = match bat.get(sector.bat_index as usize) {
+            Some(entry) => (*entry).0,
```
```suggestion
            Some(entry) => entry.0,
```

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED_V095
+                | vhdx_bat::PAYLOAD_BLOCK_ZERO => {}
+                vhdx_bat::PAYLOAD_BLOCK_FULLY_PRESENT => {
+                    // should we put lock here or in the calling function?
```
You can remove this comment, the fact that this function expect a mutable reference onto the file means that the caller will have to make sure there's no concurrency.

sboeuf commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,235 @@
+                        | (vhdx_bat::PAYLOAD_BLOCK_FULLY_PRESENT & vhdx_bat::BAT_STATE_BIT_MASK);
+                    bat.insert(sector.bat_index as usize, BatEntry(updated_bat_entry));
+                    // TODO: propagate
+                    let _ = BatEntry::write_bat_entries(f, bat_offset, bat.clone());
```
Definitely needs some `map_err()?` in here.

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+
+        #[sorted]
+        match self {
+            InvalidBlockSize => write!(f, "invalid block size count"),
+            InvalidEntryCount => write!(f, "invalid metadata entry count"),
+            InvalidLogicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidMetadataItem => write!(f, "invalid metadata ID"),
+            InvalidMetadataLength => write!(f, "invalid metadata length"),
+            InvalidMetadataSign => write!(f, "metadata sign doesn't match"),
+            InvalidPhysicalSectorSize => write!(f, "invalid logical sector size"),
+            InvalidValue => write!(f, "invalid value"),
+            MissingMetadata => write!(f, "not all required metadata found"),
+            ReadMetadata(e) => write!(f, "failed to read metadata headers: {}", e),
+            UnsupportedFlag => write!(f, "this implementation support this metadata flag"),
```
Could you please put a capital at the beginning of each sentence?

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+}
+
+struct MetadataTableHeader {
+    _signature: u64,
```
```suggestion
    signature: u64,
```

fazlamehrab replied:
We get a warning because we read the signature from the file, make sure if it is the expected one, and store the value in the structure but never use it afterward.

fazlamehrab replied:
With Default(), we do not get the warning!

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+            return Err(Error::InvalidMetadataSign);
+        }
+
+        let _reserved = LittleEndian::read_u16(&buffer[8..10]);
```
No need to read the reserved field.

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+    pub total_sectors: u64,
+}
+
+struct MetadataTableHeader {
```
Derive `Default`, this way when you need to initialize the structure, you don't need to bother about your reserved fields.

fazlamehrab replied:
Please see if my latest implementation with Default works.

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+    }
+}
+
+pub struct MetadataTableEntry {
```
Same here, derive `Default`.

sboeuf commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,322 @@
+            offset: LittleEndian::read_u32(&buffer[16..20]),
+            length: LittleEndian::read_u32(&buffer[20..24]),
+            flag_bits: LittleEndian::read_u32(&buffer[24..28]),
+            _reserved: LittleEndian::read_u32(&buffer[28..32]),
```
Don't initialize `_reserved`, just use the default implementation of `MetadataTableEntry`.

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+
+        #[sorted]
+        match self {
+            CalculateChecksum => write!(f, "Failed to calculate checksum"),
+            DuplicateBATEntry => write!(f, "BAT entry is not unique"),
+            DuplicateMDREntry => write!(f, "Metadata region entry is not unique"),
+            InvalidChecksum(section) => write!(f, "{} checksum doesn't match", section),
+            InvalidEntryCount => write!(f, "invalid entry cound"),
+            InvalidHeaderSign => write!(f, "not a valid VHDx header"),
+            InvalidRegionSign => write!(f, "not a valid VHDx region"),
+            InvalidUuid => write!(f, "couldn't parse Uuid for region entry"),
+            InvalidVHDXSign => write!(f, "not a VHDx file"),
+            NoValidHeader => write!(f, "No valid header found"),
+            ReadChecksum => write!(f, "cannot read checksum"),
+            ReadFileTypeIndetifier(e) => write!(f, "failed to read File Type Identifier: {}", e),
+            ReadHeader(e) => write!(f, "failed to read headers: {}", e),
+            ReadRegionTableEntry(e) => write!(f, "failed to read region table entry: {}", e),
+            ReadRegionTableHeader(e) => write!(f, "failed to read region table header: {}", e),
+            RegionEntryCollectionFailed => write!(f, "failed to read region entries"),
+            RegionOverlap => write!(f, "overlapping regions found"),
+            ReservedIsNonZero => write!(f, "reserved region has non-zero value"),
+            UnrecognizedRegionEntry => write!(f, "we do not recongize this entry"),
+            WriteHeader(e) => write!(f, "failed to write header: {}", e),
```
Make sure you're consistent across error messages, including using a capital letter at the beginning.

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+impl Header {
+    /// Reads the Header structure from a reference VHDx file
+    pub fn new(f: &mut File, start: u64) -> Result<Header> {
+        // Read the whole header in to a buffer. We will need if for calculating checksum
```
```suggestion
        // Read the whole header in to a buffer. We will need it for calculating checksum
```

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+        start: u64,
+    ) -> Result<Header> {
+        let mut buffer = [0u8; HEADER_SIZE as usize];
+        let mut new_header: Header;
+        let data_write_guid;
```
Not really idiomatic to Rust. You should declare these variables when they get filled with data.

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+        let file_type_identifier: FileTypeIdentifier = FileTypeIdentifier::new(f)?;
+        let header_1 = Header::new(f, HEADER1_START);
+        let header_2 = Header::new(f, HEADER2_START);
+        let (header_1, header_2) = match VhdxHeader::update_headers(f, header_1, header_2, guid) {
+            Ok(h) => h,
+            Err(e) => return Err(e),
+        };
```
```suggestion
let (header_1, header_2) = VhdxHeader::update_headers(f, header_1, header_2, guid)?;
```

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+const FILE_START: u64 = 0; //The first element
+const HEADER1_START: u64 = 64 * 1024; //header 1 start in Bytes
+const HEADER2_START: u64 = 128 * 1024; //header 2 start in Bytes
+const REGION1_START: u64 = 192 * 1024; // Region 1 start in Bytes
+const REGION2_START: u64 = 256 * 1024; // Region 2 start in Bytes
```
```suggestion
const REGION_TABLE_1_START: u64 = 192 * 1024; // Region 1 start in Bytes
const REGION_TABLE_2_START: u64 = 256 * 1024; // Region 2 start in Bytes
```

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+const REGION_SIGN: u32 = 0x6967_6572; // "regi"
+
+const FILE_START: u64 = 0; //The first element
+const HEADER1_START: u64 = 64 * 1024; //header 1 start in Bytes
+const HEADER2_START: u64 = 128 * 1024; //header 2 start in Bytes
```
```suggestion
const HEADER_1_START: u64 = 64 * 1024; //header 1 start in Bytes
const HEADER_2_START: u64 = 128 * 1024; //header 2 start in Bytes
```

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+    ) -> Result<(HeaderNo, Header)> {
+        let mut header1_seq_num: u64 = 0;
+        let mut header2_seq_num: u64 = 0;
+        let mut no_valid_header: bool = true;
```
Nit: It'd be more natural to call it `valid_hdr_found`, therefore inverting the logic of the boolean.

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+    /// _update_header() takes two headers and update the noncurrent header with
```
```suggestion
    /// This function takes two headers and update the noncurrent header with
```

sboeuf commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,524 @@
+    /// _update_header() takes two headers and update the noncurrent header with
+    /// the corrent one. Returns both headers as a tuple sequenced as it received
+    /// them in the parameter list.
+
```
Remove this blank line.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,320 @@
+    offset: u32,
+    length: u32,
+    flag_bits: u32,
+    _reserved: u32,
```
Since you read each field in individually (a poor choice imho) you don't need this field.

sboeuf replied:
Please provide the alternative you're thinking about so that @fazlamehrab can implement it.

rbradford replied:
To be clear: CH only supports LE systems. So if you keep the reserved fields, mark the struct as `#[repr(packed)]` then you can just read the whole struct from the file in one go.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,320 @@
+#[derive(Default, Debug)]
+struct MetadataTableHeader {
+    signature: u64,
+    _reserved: u16,
```
Ditto.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,320 @@
+    signature: u64,
+    _reserved: u16,
+    entry_count: u16,
+    _reserved2: [u8; 20],
```
Ditto.

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,320 @@
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
```
Please use `thiserror` instead, search the code for `#[error(...)]`

rbradford commented on vhdx/src/vhdx_io.rs:
```diff
@@ -0,0 +1,231 @@
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
```
Ditto.

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,521 @@
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
```
Ditto.

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,521 @@
+    pub file_write_guid: u128,
+    pub data_write_guid: u128,
+
+    pub log_guid: u128,
```
Why this whitespace?

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,521 @@
+    pub checksum: u32,
+    pub entry_count: u32,
+    pub reserved: u32,
+}
```
Ditto.

fazlamehrab replied:
According to the spec, reserved fields MUST be zero. So, we need to check them to see if they are consistent with the spec.

rbradford commented on vhdx/src/vhdx_header.rs:
```diff
@@ -0,0 +1,521 @@
+    pub entry_count: u32,
+    pub reserved: u32,
+}
+impl RegionTableHeader {
```
Missing a whitespace.

rbradford commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,116 @@
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
```
Ditto.

rbradford commented:
@fazlamehrab I pushed a commit that enhanced the fuzzer so that it had a greater chance of finding an issue. It already found one:

```
thread '<unnamed>' panicked at 'attempt to subtract with overflow', /home/rob/src/cloud-hypervisor-fuzz/vhdx/src/vhdx_metadata.rs:154:47
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
==1655896== ERROR: libFuzzer: deadly signal
```

sboeuf commented:
@fazlamehrab I have a commit [here](https://github.com/sboeuf/cloud-hypervisor-1/commit/bc51d7382efceca8aafaf49949c04b53c66a7867) to show you how you can handle converting what you have in memory (buffer) to a structure. Using `#[repr(C)]` is mandatory to make sure the fields in the structure are ordered correctly.

fazlamehrab commented on vhdx/src/vhdx.rs:
```diff
@@ -0,0 +1,236 @@
+    /// Wrapper function to satisfy the Read trait implementation for VHDx disk. Converts the offest to secotor index and buffer length to sector count.
+    fn read(&mut self, buf: &mut [u8]) -> std::result::Result<usize, std::io::Error> {
+        let sector_count =
+            div_round_up!(buf.len() as u64, self.disk_spec.logical_sector_size as u64);
```
Decided not to proceed with this approach after discussing with Sebastien.

fazlamehrab commented on vhdx/src/vhdx_bat.rs:
```diff
@@ -0,0 +1,126 @@
+pub const PAYLOAD_BLOCK_UNMAPPED: u64 = 3;
+pub const PAYLOAD_BLOCK_UNMAPPED_V095: u64 = 5;
+pub const PAYLOAD_BLOCK_FULLY_PRESENT: u64 = 6;
+pub const PAYLOAD_BLOCK_PARTIALLY_PRESENT: u64 = 7;
```
We will take care of this in the next PR.

fazlamehrab commented:
Hi @russell-islam @likebreath @rbradford @sboeuf , I have finished addressing all the comments from your reviews. Therefore, it would be great if you please do another pass on the latest state of the PR so that we can have it ready to be accepted. Thanks!

rbradford commented on block_util/src/lib.rs:
```diff
@@ -623,6 +626,9 @@ pub fn detect_image_type(f: &mut File) -> std::io::Result<ImageType> {
         ImageType::Qcow2
     } else if vhd::is_fixed_vhd(f)? {
         ImageType::FixedVhd
+    //} else if vhdx::is_vhdx(f)? {
```
Commented out code.

rbradford commented on fuzz/fuzz_targets/qcow.rs:
```diff
@@ -11,33 +11,31 @@ use std::io::{self, Cursor, Read, Seek, SeekFrom, Write};
-            let _ = qcow.write_all(&value.to_le_bytes());
+        if qcow.seek(SeekFrom::Start(0)).is_ok() {
+            let mut offset = 0;
+            while offset < bytes.len() {
```
You should drop the qcow change and squash into the other fuzzer commit

rbradford commented on vhdx/src/vhdx_metadata.rs:
```diff
@@ -0,0 +1,330 @@
+    signature: u64,
+    reserved: u16,
+    entry_count: u16,
+    _reserved2: [u8; 20], // helps get actual Metadata Table Header size
```
You don't need this comment since you're going the packed approach.

sboeuf commented:
@fazlamehrab I've pushed an extra commit with some suggestions for the integration tests. If you agree with it, please squash it into the actual commit introducing the integration tests.
I also looked at the dynamic support in `vhdx_io.rs`, the code looks good to me.
